### DISCUSSION TITLE
Implementing symmetries with phases and antiunitary operators

### DISCUSSION THREAD

**Discussion by msoldi** [QuSpin/QuSpin#497] [Q&A]
Hi!

I would like to implement symmetries that involve state-dependent phases in spinful_fermion_basis_general, including antiunitary operations such as time-reversal symmetry (TRS). 

For example, in a chain of spinful fermions I may want to apply a mirror symmetry that acts on the fermionic annihilation operators as
c_{r, up} --->   c_{-r, down}
c_{r, down} --->  - c_{-r, up}
with 'r' indexing the fermion position and 'up/down' indicating the spin direction. I would like to implement this symmetry in the fermion basis, to reduce the size of the Hilbert space correctly.

As an explicit example in the case of mirror symmetry, if I have two spinful lattice sites (r=0, 1) indexed as
`# [0, 1, 2, 3] == [r=0 and spin up, r=1 and spin up, r=0 and spin down, r=1 and spin down ]`
I would like to be able to write something of the type
`M = [-3, -2, 1, 0]`
`basis = spinful_fermion_basis_general(2, simple_symm=False, M_block=(M, 1j))`
where I can select the +1j or -1j sector of the mirror symmetry, analogous to what one does with 'standard' symmetries such as translation. If I try to implement the lines above, I get the error
`ValueError: map must be a one-to-one site mapping.`
Is there a way I can implement these symmetries inside the basis using symmetry blocks?
Thank you for any help!

**Comment by JiahaoYao** [comment]
Hi @msoldi , i am not very familiar with this symmetry and you can refer more to this [page](http://weinbe58.github.io/QuSpin/generated/quspin.basis.spinful_fermion_basis_general.html). 

> **blocks: optional
keyword arguments which pass the symmetry generator arrays. For instance:
basis(...,kxblock=(Q,q),...)
The keys of the symmetry sector, e.g. kxblock, can be chosen arbitrarily by the user. The values are tuples where the first entry contains the symmetry transformation 𝑄 acting on the lattice sites (see class example), and the second entry is an integer 𝑞 to label the symmetry sector.


I can reproduce your error: 

```python
>>> M = [-3, -2, 1, 0]
>>> from quspin.basis import spinful_fermion_basis_general
>>> basis = spinful_fermion_basis_general(2, simple_symm=False, M_block=(M, 1j))
Traceback (most recent call last):
  File "/python3.7/site-packages/quspin/basis/basis_general/base_general.py", line 23, in process_map
    raise ValueError("map must be a one-to-one site mapping.")
ValueError: map must be a one-to-one site mapping.
```

So, you might need to change your `M` to the rearrange of `[1, 2, 3, 4]`. For example, `M = [3, 2, 1, 0]`. 

```python 
>>> M = [3, 2, 1, 0]
>>> basis = spinful_fermion_basis_general(2, simple_symm=False, M_block=(M, 1j))
Traceback (most recent call last):
  File "python3.7/site-packages/numpy/core/_asarray.py", line 83, in asarray
    return array(a, dtype, copy=False, order=order)
TypeError: can't convert complex to int
```

In the end, it will be 
```python 
>>> basis = spinful_fermion_basis_general(2, simple_symm=False, M_block=(M, 1))
>>> basis
<quspin.basis.basis_general.fermion.spinful_fermion_basis_general object at 0x7f97ba2ff290>
```

This might not solve your issue, and there are two things: when you do `kxblock=(Q,q)` in the symmetry, i) Q needs to be the array rearragement of `[0, 1, ...., N]` and ii) `q` needs to be integer.

**Reply by msoldi** [reply]
Hi @JiahaoYao 

Thank you for your answer, I agree with you: when you define symmetries with SYMMETRY_block = (Q, q), where Q is equal to some array rearrangement and q is integer, there are no issues at all. However, to implement some of the symmetries of my problem this is not enough, since I also need to include additional phases appearing from the symmetry action (as in the mirror symmetry of my example, I need to add the phase '-1' for each spin down electron). For this, describing Q as an arrangement of integers is not enough when I also want to take into account the phase acquired by the basis states. My specific example is simplified, and there are ways of expressing the mirror symmetry I wrote above without having to include the '-1' sign, but I am working with more general cases where it is not easy to avoid the introduction of phases in the action of symmetries.

I would like to know if there is some general strategy with which I can achieve what the SYMMETRY_block=(Q, q) option in  spinful_fermion_basis_general does, which may be applicable to the case of symmetries described not only by a rearrangement of the indexes of the fermion basis but also by an index-dependent phase.

Thank you for any help

**Reply by JiahaoYao** [reply]
Hi @msoldi , yes I agree with you! I am not very sure whether the current version has already supported this feature or there is some way to get around this. It would be interesting if @mgbukov can comment on this.

**Reply by adithi16** [reply]
Hi, 
I am stuck with a similar issue to implement a general symmetry which also involves spin operators and not just a rearrangement of lattice sites. Any update on this? 
Thank you!

**Reply by mgbukov** [reply]
can you give an example?

**Reply by adithi16** [reply]
> can you give an example?

Thank you, you replied to my issue at https://github.com/QuSpin/QuSpin/issues/609 :)
